### ISSUE TITLE
How to handle `scan` primitive with implicit arrays?

### ISSUE THREAD

**Issue by giganttheo** [davisyoshida/qax#5]
Hi, I am using the `jax-lorax` library based on qax and I try to use `flax.linen.scan` in order to reduce memory and time usage during training. However it seems that the scan primitive is not yet handled by qax's implicit arrays. I checked the code for primitive handlers ( https://github.com/davisyoshida/qax/blob/master/qax/implicit/implicit_array.py ) and I have no idea how to implement that by myself.

Are you planning on supporting the scan primitive in this library?

**Comment by davisyoshida** [comment]
Yeah it's a bit tricky, I'll add it now.

**Comment by davisyoshida** [comment]
@giganttheo Okay so there turned out to be more edge cases than I thought of... To unblock you I implemented support for scanning over `ImplicitArray` instances, but if they're used in the _carry_ they'll get materialized. Importantly, you'll also need to use `lorax` at HEAD to get the expected results, since there was a bug leading to incorrect scaling of the LoRAs in the case of conv layers (i.e. LoRAs with more than two dimensions). Here's a code sample of a scan:

```python
from functools import partial

import jax
import jax.numpy as jnp

import lorax

def main():
    batch_size = 2
    w = jnp.ones((batch_size, 5, 5))
    a = jax.random.normal(jax.random.PRNGKey(0), (batch_size, 3, 5))
    b = jax.random.normal(jax.random.PRNGKey(1), (batch_size, 5, 3))
    lora = lorax.LoraWeight(w=w, a=a, b=b)

    x = jax.random.normal(jax.random.PRNGKey(2), (5,))

    def body(carry, w):
        output = carry @ w
        return output, None

    fn = lorax.lora(partial(jax.lax.scan, body))
    expected, _ = fn(x, lora.materialize())
    actual, _ = fn(x, lora)

    assert jnp.allclose(expected, actual)


if __name__ == '__main__':
    main()
```

Let me know if you run into any problems.

**Comment by giganttheo** [comment]
Hi @davisyoshida thank you for the quick response and update of the library. It works fine on minimal examples like the one you run in your test files, as well as with flax's `nn.scan` function. 

However for some reason it does not work for my complete codebase and I am still investigating why.

This is the error that is thrown when using a lora-ed model. The same code works fine without lora.

```
  File "/mnt/beegfs/projects/graph-transformer/aga_transformers/aga_transformers/models/t5/modeling_t5_augmented_efficient.py", line 1537, in __call__
    layer_outputs, _ = nn.scan(FlaxT5LayerCollection,#remat(FlaxT5LayerCollection, static_argnums=(6, 7, 8)),
  File "/mnt/beegfs/home/gigant/miniconda3/envs/train-jax/lib/python3.8/site-packages/flax/core/axes_scan.py", line 163, in scan_fn
    c, ys = lax.scan(
  File "/mnt/beegfs/projects/graph-transformer/qax/qax/implicit/implicit_array.py", line 391, in process_primitive
    outs = _default_handlers[primitive.name](primitive, *vals, params=params)
  File "/mnt/beegfs/projects/graph-transformer/qax/qax/implicit/implicit_array.py", line 504, in _handle_scan
    sliced_xs = jax.tree_map(
  File "/mnt/beegfs/projects/graph-transformer/qax/qax/implicit/implicit_array.py", line 505, in <lambda>
    partial(jax.eval_shape, lambda x: x[0]),
  File "/mnt/beegfs/home/gigant/miniconda3/envs/train-jax/lib/python3.8/site-packages/jax/_src/numpy/array_methods.py", line 791, in op
    return getattr(self.aval, f"_{name}")(self, *args)
  File "/mnt/beegfs/home/gigant/miniconda3/envs/train-jax/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 4143, in _rewriting_take
    return _gather(arr, treedef, static_idx, dynamic_idx, indices_are_sorted,
  File "/mnt/beegfs/home/gigant/miniconda3/envs/train-jax/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 4152, in _gather
    indexer = _index_to_gather(shape(arr), idx)  # shared with _scatter_update
  File "/mnt/beegfs/home/gigant/miniconda3/envs/train-jax/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 4253, in _index_to_gather
    idx = _canonicalize_tuple_index(len(x_shape), idx)
  File "/mnt/beegfs/home/gigant/miniconda3/envs/train-jax/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 4586, in _canonicalize_tuple_index
    raise IndexError(
IndexError: Too many indices for array: 1 non-None/Ellipsis indices for dim 0.  File "/mnt/beegfs/projects/graph-transformer/aga_transformers/aga_transformers/models/t5/modeling_t5_augmented_efficient.py", line 1537, in __call__
    layer_outputs, _ = nn.scan(FlaxT5LayerCollection,#remat(FlaxT5LayerCollection, static_argnums=(6, 7, 8)),
  File "/mnt/beegfs/home/gigant/miniconda3/envs/train-jax/lib/python3.8/site-packages/flax/core/axes_scan.py", line 163, in scan_fn
    c, ys = lax.scan(
  File "/mnt/beegfs/projects/graph-transformer/qax/qax/implicit/implicit_array.py", line 391, in process_primitive
    outs = _default_handlers[primitive.name](primitive, *vals, params=params)
  File "/mnt/beegfs/projects/graph-transformer/qax/qax/implicit/implicit_array.py", line 504, in _handle_scan
    sliced_xs = jax.tree_map(
  File "/mnt/beegfs/projects/graph-transformer/qax/qax/implicit/implicit_array.py", line 505, in <lambda>
    partial(jax.eval_shape, lambda x: x[0]),
  File "/mnt/beegfs/home/gigant/miniconda3/envs/train-jax/lib/python3.8/site-packages/jax/_src/numpy/array_methods.py", line 791, in op
    return getattr(self.aval, f"_{name}")(self, *args)
  File "/mnt/beegfs/home/gigant/miniconda3/envs/train-jax/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 4143, in _rewriting_take
    return _gather(arr, treedef, static_idx, dynamic_idx, indices_are_sorted,
  File "/mnt/beegfs/home/gigant/miniconda3/envs/train-jax/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 4152, in _gather
    indexer = _index_to_gather(shape(arr), idx)  # shared with _scatter_update
  File "/mnt/beegfs/home/gigant/miniconda3/envs/train-jax/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 4253, in _index_to_gather
    idx = _canonicalize_tuple_index(len(x_shape), idx)
  File "/mnt/beegfs/home/gigant/miniconda3/envs/train-jax/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 4586, in _canonicalize_tuple_index
    raise IndexError(
IndexError: Too many indices for array: 1 non-None/Ellipsis indices for dim 0.
```

Just so you know, I changed a little bit the code to create a lora version of parameters in [`init_lora`](https://github.com/davisyoshida/lorax/blob/master/lorax/helpers.py#L20) to create the equivalent of stacked 2d-shaped parameters:

```python
if len(param.shape) == 3:
    #case for scanned fn:
    layer_dim, b_dim, a_dim = param.shape
    b = jnp.zeros((layer_dim, b_dim, spec_val), dtype=dtype)
    a = jax.random.normal(next(key_it), (layer_dim, spec_val, a_dim), dtype=dtype) * stddev
    return LoraWeight(w=param, a=a, b=b, alpha=alpha)
```

I don't understand where the issue comes from but my understanding is that at some point the scanning process goes over a scalar lora weight. I checked all the weights that are transformer into LoraWeights and all of them are the correct shape (ie with at least 3 dimensions). I tried to reproduce the issue with a minimal example but I don't find how to get the same error. I will keep trying and I'll let you know when I find the issue. If you have any idea what might be the issue it would be very helpful.

On another note, I think that the dependencies of qax on other libraries might be altered. In the current code you are using `jax.extend.linear_util` that seems to be introduced in `jax` `0.4.16`, I don't know if the package content is the same as the older `jax.linear_util` but either way it might break qax's compatibility with jax<0.4.16.

**Comment by davisyoshida** [comment]
@giganttheo Hmm yeah I can't repro it even with that code you're using to init the `LoraWeight` instances, and it seems to also be compatible with the example code Flax gives for a scannable network. Can you run the following on your parameter tree and post it here?

```python
from pprint import pprint
pprint(jax.tree_map(
    lambda x: jax.tree_map(jnp.shape, x) if isinstance(x, LoraWeight) else None,
    params,
    is_leaf=lambda x: isinstance(x, LoraWeight)
))
```

If possible, you could also send me your model definition code to look at. Is there any chance you've accidentally made alpha a parameter which it's trying to scan? Does it definitely work when you run it without LoRAs?

> I think that the dependencies of qax on other libraries might be altered

Ah good catch, yeah I had to switch that since they've deprecated `jax.linear_util`. I'll update the dependency and cut a release once we get this figured out.

**Comment by giganttheo** [comment]
I found the bug @davisyoshida

It has to do with the variable broadcasting in `nn.scan`:

The following code doesn't work:

```python
import jax
import jax.numpy as jnp
import flax.linen as nn

import lorax
import qax

class Body(nn.Module):
  @nn.compact
  def __call__(self, x, attention_mask):
      return nn.Dense(6, use_bias=False)(x) * attention_mask, None

class ScanModel(nn.Module):
  @nn.compact
  def __call__(self, x, attention_mask):
    return nn.scan(Body, length=10, in_axes=(nn.broadcast, ), variable_axes={"params": 0}, split_rngs={"params": True})()(x, attention_mask)[0]

X_train = jnp.arange(24).reshape((4, 6)).astype("float32")
attention_mask = jnp.ones((4, 6), dtype="bool")
variables = ScanModel().init(jax.random.PRNGKey(0), X_train, attention_mask)
lora_params=jax.tree_util.tree_map(lambda w: lorax.LoraWeight(w=w, a=jnp.ones((w.shape[0], 1, w.shape[-1])), b=jnp.ones((w.shape[0], w.shape[-2], 1)), alpha=1), variables)
vanilla_params=jax.tree_util.tree_map(lambda w: lorax.LoraWeight(w=w, a=jnp.ones((w.shape[0], 1, w.shape[-1])), b=jnp.ones((w.shape[0], w.shape[-2], 1)), alpha=1).materialize(), variables)

expected_output = ScanModel().apply(vanilla_params, X_train, attention_mask)
lora_output = qax.use_implicit_args(ScanModel().apply)(lora_params, X_train, attention_mask)

assert jnp.allclose(expected_output, lora_output)
```

The `expected_output` is computed as expected but the script fails when computing the `lora_output`.

With a slight modification in the code it can work:

```python
[...]
    return nn.scan(Body, length=10, in_axes=(0, ), variable_axes={"params": 0}, split_rngs={"params": True})()(x, attention_mask)[0]

[...]
attention_mask = jnp.ones((10, 4, 6), dtype="bool")

```

The only difference between both is the value for argument `in_axes` in `nn.scan`, and the `attention_mask` shape.
With `nn.broadcast` as `in_axes`, the scan inputs after the first one (which is the carry), are supposed to be broadcast for each loop pattern ([comment related to this](https://github.com/google/flax/discussions/2127)), while when manually broadcast this is working even with the Lora weights.

The weird thing is that the broadcast argument is not using Lora weights, nor changed during scanning.

I hope that helps if you intend to fix this scenario.
In my case I can fix the issues and make it work as intended by avoiding the use of `nn.broadcast`

**Comment by davisyoshida** [comment]
Thanks a ton for the short repro! I'll take a look when I get a chance to see what I can do.

**Comment by davisyoshida** [comment]
@gigantheo I believe I tracked down the bug, it should be fixed in 0.3.1 which I've just put on pypi. Thanks again!

**Comment by giganttheo** [comment]
@davisyoshida awesome thank you. After testing it in my code I can confirm that it works well for me :)
